File: src/support_sphere_py/src/support_sphere/models/public/resource.py
from typing import Optional

from support_sphere.models.base import BasePublicSchemaModel
from sqlmodel import Field, Relationship


class Resource(BasePublicSchemaModel, table=True):
    """
    Represents a resource entity in the 'public' schema under the 'resources' table.
    Resources represent items or assets that have attributes like quantity, description, and tags.

    Attributes
    ----------
    id : int, optional
        The unique identifier for the resource. This is the primary key.
    resource_type : int, optional
        A foreign key referring to the 'resource_types' table that categorizes the resource.
    name : str, optional
        The name of the resource. This field is required.
    description : str, optional
        A description of the resource, providing details about its use or nature.
    notes : str, optional
        Additional notes or comments about the resource.
    qty_needed : int, optional
        The quantity of the resource that is needed. Default is 0.
    qty_available : int, optional
        The quantity of the resource that is available. Default is 0.
    resource_tags : list[ResourceTag]
        Defines a one-to-many relationship with the `ResourceTag` model, allowing multiple tags to be associated with a resource.
    user_resources : list[UserResource]
        Defines a one-to-many relationship with the `UserResource` model, associating user with specific resources.
    """

    __tablename__ = "resources"

    id: int | None = Field(primary_key=True)
    resource_type_id: int|None = Field(foreign_key="public.resource_types.id", nullable=False)
    name: str | None = Field(nullable=False)
    description: str | None = Field(nullable=True)
    notes: str | None = Field(nullable=False)

    qty_needed: int|None = Field(nullable=False, default=0)
    qty_available: int|None = Field(nullable=False, default=0)

    resource_tags: list["ResourceTag"] = Relationship(back_populates="resources", cascade_delete=False)
    user_resources: list["UserResource"] = Relationship(back_populates="resource", cascade_delete=False)
    resource_type: Optional["ResourceType"] = Relationship(back_populates="resources", cascade_delete=False)

File: src/support_sphere/lib/data/models/resource.dart
import 'package:equatable/equatable.dart';
import 'package:support_sphere/data/models/generated_classes.dart';

class Resource extends Equatable {
  const Resource({
    required this.id,
    required this.name,
    required this.resourceType,
    required this.resourceCv,
    this.notes,
    this.description = '',
    this.qtyNeeded = 0,
    this.qtyAvailable = 0,
    this.userQuantity = 0,
    this.sharingScope = '',
    this.sharingScopeEmergency = '',
  });

  final String id;
  final String name;
  final String? description;
  final String? notes;
  final String? sharingScope;
  final String? sharingScopeEmergency;
  final int qtyNeeded;
  final int qtyAvailable;
  final int userQuantity;
  final ResourceTypes resourceType;
  final ResourcesCv resourceCv;

  @override
  List<Object?> get props => [id, name, description];

  static Resource fromJson(Map<String, dynamic> json) {
    var resourceTypesJson = json['resource_types'];
    var resourcesCvJson = json['resources_cv'];
    var userQuantity = 0;
    if (json['user_resources'].length > 0) {
      userQuantity = json['user_resources']
          .map((userResource) => userResource['quantity'])
          .reduce((a, b) => a + b);
    }
    var neededQuantity = json['qty_needed'] - userQuantity;
    return Resource(
      id: resourcesCvJson['id'],
      name: resourcesCvJson['name'],
      description: resourcesCvJson['description'],
      resourceType: ResourceTypes.fromJson(resourceTypesJson),
      resourceCv: ResourcesCv.fromJson(resourcesCvJson),
      notes: json['notes'],
      sharingScope: json['sharing_scope'],
      sharingScopeEmergency: json['sharing_scope_emergency'],
      qtyNeeded: neededQuantity < 0 ? 0 : neededQuantity,
      qtyAvailable: json['qty_available'] + userQuantity,
    );
  }

  Resource copyWith({
    String? id,
    String? name,
    String? description,
    String? notes,
    String? sharingScope,
    String? sharingScopeEmergency,
    int? qtyNeeded,
    int? qtyAvailable,
    int? userQuantity,
    ResourceTypes? resourceType,
    ResourcesCv? resourceCv,
  }) {
    return Resource(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      resourceType: resourceType ?? this.resourceType,
      resourceCv: resourceCv ?? this.resourceCv,
      notes: notes ?? this.notes,
      sharingScope: sharingScope ?? this.sharingScope,
      sharingScopeEmergency:
          sharingScopeEmergency ?? this.sharingScopeEmergency,
      userQuantity: userQuantity ?? this.userQuantity,
      qtyNeeded: qtyNeeded ?? this.qtyNeeded,
      qtyAvailable: qtyAvailable ?? this.qtyAvailable,
    );
  }
}
